spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/reactive/PubSubReactiveFactory.java

```diff
@@ -77,116 +86,62 @@ public PubSubReactiveFactory(PubSubSubscriberOperations subscriberOperations, Sc
-
-		protected final FluxSink<AcknowledgeablePubsubMessage> sink;
-
-		PubSubPullTask(String subscriptionName, FluxSink<AcknowledgeablePubsubMessage> sink) {
-			this.subscriptionName = subscriptionName;
-			this.sink = sink;
-		}
-
-		/**
-		 * Retrieve up to a specified number of messages, sending them to the subscription.
-		 * @param demand maximum number of messages to retrieve
-		 * @param block whether to wait for the first message to become available
-		 * @return number of messages retrieved
-		 */
-		protected int pullToSink(int demand, boolean block) {
-
-			List<AcknowledgeablePubsubMessage> messages =
-					PubSubReactiveFactory.this.subscriberOperations.pull(this.subscriptionName, demand, !block);
-
-			if (!this.sink.isCancelled()) {
-				messages.forEach(sink::next);
-			}
-
-			return messages.size();
-		}
+	private void pollingPull(String subscriptionName, long pollingPeriodMs,
+			FluxSink<AcknowledgeablePubsubMessage> sink) {
+		Disposable disposable = Flux
+				.interval(Duration.ZERO, Duration.ofMillis(pollingPeriodMs), scheduler)
+				.flatMap(ignore -> pullAll(subscriptionName))
+				.subscribe(sink::next, sink::error);
 
+		sink.onDispose(disposable);
 	}
 
-	/**
-	 * Runnable task issuing blocking Pub/Sub Pull requests until the specified number of
-	 * messages has been retrieved.
-	 */
-	private class BlockingLimitedDemandPullTask extends PubSubPullTask {
+	private Flux<AcknowledgeablePubsubMessage> pullAll(String subscriptionName) {
+		CompletableFuture<List<AcknowledgeablePubsubMessage>> pullResponseFuture = this.subscriberOperations
+				.pullAsync(subscriptionName, Integer.MAX_VALUE, true).completable();
 
-		private final long initialDemand;
-
-		BlockingLimitedDemandPullTask(String subscriptionName, long initialDemand, FluxSink<AcknowledgeablePubsubMessage> sink) {
-			super(subscriptionName, sink);
-			this.initialDemand = initialDemand;
-		}
-
-		@Override
-		public void run() {
-			try {
-				long demand = this.initialDemand;
-				List<AcknowledgeablePubsubMessage> messages;
+		return Mono.fromFuture(pullResponseFuture).flatMapMany(Flux::fromIterable);
+	}
 
-				while (demand > 0 && !this.sink.isCancelled()) {
-					try {
-						int intDemand = demand > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) demand;
-						demand -= pullToSink(intDemand, true);
+	private void backpressurePull(String subscriptionName, long numRequested,
+			FluxSink<AcknowledgeablePubsubMessage> sink) {
+		int intDemand = numRequested > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) numRequested;
+		this.subscriberOperations.pullAsync(subscriptionName, intDemand, false).addCallback(
+				messages -> {
+					if (!sink.isCancelled()) {
+						messages.forEach(sink::next);
 					}
-					catch (DeadlineExceededException e) {
+					if (!sink.isCancelled()) {
+						long numToPull = numRequested - messages.size();
+						if (numToPull > 0) {
+							backpressurePull(subscriptionName, numToPull, sink);
```

Comment:
Indeed, this method seems to be recursive, but it is not because the method calls itself within the context of a different thread than itself was executed. There is no thread stack build up.

Regarding your second point: is this actually a problem? PubSub should be able to handle it just fine. It is made for concurrent outstanding calls. It even recommends to do concurrent calls, to be able to get the highest throughput. It's just a bit unpredictable how many outstanding calls there can be at any given time.

If this is a problem then I think we can change this so that it is guaranteed that there is only on process retrieving the messages at any give time. When a new request comes in for new messages, it will then pass this on to this process and it will add it to the existing number of other messages that it already tries to pull. There is probably an elegant way of doing this in reactor...

Reply:
Look fine to me. Good point regarding two different threads being involved which avoids the stack buildup.